packages/dev-middleware/src/inspector-proxy/DeviceMessageMiddleware.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow strict-local
+ * @format
+ */
+
+import type {DebuggerInfo} from './Device';
+import type {MessageFromDevice, MessageToDevice, Page} from './types';
+import type WS from 'ws';
+
+type ExposedDeviceInfo = $ReadOnly<{
+  appId: string,
+  id: string,
+  name: string,
+  socket: WS,
+}>;
+
+type ExposedDebuggerInfo = $ReadOnly<{
+  socket: $ElementType<DebuggerInfo, 'socket'>,
```

Comment:
We can't expose the raw socket to middleware. If the middleware needs the ability to send CDP messages, then `CreateDeviceMessageMiddlewareFn` should include a method to do that as a parameter.

Reply:
Makes sense, in our implementation I added the `sendToDebugger` and sendToDevice` methods anyways. I think we can move them to the `debuggerInfo`/`deviceInfo`, without exposing the socket directly. 

e.g. 
- `debuggerInfo.send = (message: DebuggerEvent | DebuggerMessage) => void`
- `deviceInfo.send = (message: DeviceEvent | DeviceMessage) => void`

Would that solve your point here? Not 100% sure yet about the typing, but you get the point.

Reply:
✅ I replaced the sockets with `sendMessage` methods.

Here are our custom handlers, and a quick check if this should work or not.

- [`NetworkResponse`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/NetworkResponse.ts#L35C19-L38)
  - This sends a response once `Network.getResponseBody` is requested from the debugger, should work fine.
- [`PageReload`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/PageReload.ts#L17)
  - Only replies with an acknowledge response to the debugger for  `Page.reload` requests, should work fine.
- [`VscodeDebuggerGetPossibleBreakpoints`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/VscodeDebuggerGetPossibleBreakpoints.ts#L17-L20)
  - Replies to the unsupported `Debugger.getPossibleBreakpoints` request, with an empty locations list (as vscode does not check for nullablility / the `locations: []` also required in the CDP spec). 
- [`VscodeDebuggerSetBreakpointByUrl`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/VscodeDebuggerSetBreakpointByUrl.ts)
  - Only mutates the message to avoid using unsupported way to set a breakpoint (through `Debugger.setBreakpointByUrl`)
- [`VscodeRuntimeCallFunctionOn`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/VscodeRuntimeCallFunctionOn.ts#L20-L26)
  - Replies to the debugger to avoid vscode injecting unsupported web scripts inside Hermes, which crashes Hermes / app.
- [`VscodeRuntimeGetProperties`](https://github.com/expo/expo/blob/%40bycedric/cli/inspector-device-message-middleware/packages/%40expo/cli/src/start/server/metro/debugging/deviceMiddlewares/VscodeRuntimeGetProperties.ts)
  - Only mutates the message to avoid vscode performing actions on `Symbols`, which also crashes Hermes / app.

Reply:
`Page.reload` should check the `nativePageReloads` capability flags btw.

cc @dannysu for Hermes crashes, let's write Hermes unit tests / RN integration tests around them and see if they still occur in the new backend implementation